akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -737,14 +737,23 @@ private[akka] class ShardRegion(
 
   def receiveCoordinatorMessage(msg: CoordinatorMessage): Unit = msg match {
     case HostShard(shard) =>
-      log.debug("{}: Host Shard [{}] ", typeName, shard)
-      regionByShard = regionByShard.updated(shard, self)
-      regions = regions.updated(self, regions.getOrElse(self, Set.empty) + shard)
+      if (gracefulShutdownInProgress) {
+        log.debug("{}: Ignoring Host Shard request for [{}] as region is shutting down", typeName, shard)
 
-      //Start the shard, if already started this does nothing
-      getShard(shard)
+        // if the coordinator is sending HostShard to a region that is shutting down
+        // it means that it missed the shutting down message (coordinator moved?)
+        // we want to inform it as soon as possible so it doesn't keep try to allocate the shard here
+        sendGracefulShutdownToCoordinatorIfInProgress()
```

Comment:
I think there are only two cases in which the coordinator can send `HostShard` region:

* when a RebalanceWorker timeouts, the region is not considered as shutting down on coordinator side anymore, until region tells reiterates it.
* when the coordinator moves while a region was shutting down

When that happens, the coordinator may send `HostShard`. After which, it informs the requester that the region is hosting the shard. However, that's not what we want. We want it to ignore the request and we want the coordinator to try on another region.

This is the scenario step-by-step:

1. A requests shard 1
1. coordinator decides it will be on node B
1. coordinator commands B to host it and tells A that region B hosts shard (see https://github.com/akka/akka/blob/309326fcfa164d37b2997ee2043fb85e0323d8f4/akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala#L1106-L1107)
1. B ignores the command to host
1. B reiterates to coordinator that is shutting down
1. A sends a message for shard 1 to B
1. If region B is still alive it will buffer it and ask coordinator where shard 1 is
1. Region B is hindered from shutting down while there are buffered messages
1. coordinator has now know B is shutting down and will allocate shard to another region

Region B is hindered to shutdown, but don't need to start a new shard.